[WebGPU Swift] Enable runtime flag by default when compiler is compatible

tadeuzagallo opened the pull request:
#### 2621962f0d7f295da816c4f043869d838d901173
<pre>
[WebGPU Swift] Enable runtime flag by default when compiler is compatible
<a href="https://bugs.webkit.org/show_bug.cgi?id=300935">https://bugs.webkit.org/show_bug.cgi?id=300935</a>
<a href="https://rdar.apple.com/162791460">rdar://162791460</a>

Reviewed by Mike Wyrzykowski.

Change the runtime flag that enables WebGPU Swift such that if the compiler is new
enough to support all the features we need, then it&apos;s enabled by default. It remains
possible to override it and turn it on off on any compiler version.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Capabilities.swift: Added.
(swiftCompilerSupportsWebGPU):
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::isWebGPUSwiftEnabled):

Canonical link: <a href="https://commits.webkit.org/301749@main">https://commits.webkit.org/301749@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1afc236a44386553beeae0e5a883638b58907ecf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126933 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46570 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37558 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133937 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/78513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/272d624d-6a40-4ada-b8ce-8dd5ec8cc887) 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/47190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/55101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/96590 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/78513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1ded8c6c-a3ec-4ae8-bb94-078c27c46f1c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129881 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/47190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/113533 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/77104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/eab24d02-ebdb-40be-88cd-43ed1b62f951) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/47190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77331 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/47190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32019 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136462 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/53593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/55101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105106 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54096 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109896 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104798 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26722 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50316 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28650 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19857 "Failed to checkout and rebase branch from PR 52514") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53525 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59394 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/52766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54525 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (commented):
Looks good, but we need to fix https://bugs.webkit.org/show_bug.cgi?id=300914 before merging otherwise we introduce a regression

geoffreygaren commented on Source/WebGPU/WebGPU/Capabilities.swift:
```diff
@@ -23,22 +23,10 @@
-
-#ifdef __cplusplus
+public func swiftCompilerSupportsWebGPU() -> Bool {
+    #if _compiler_version(>=6.2.3.1.1) || _compiler_version(>=6.3.0.108.2)
```
WebKit style puts #if/#else/#endif at beginning of line.

tadeuzagallo replied:
indeed, but stylecheck fails because `swift lint` complains about it... which one should we ignore?

mwyrzykowski commented on Source/WebGPU/WebGPU/HardwareCapabilities.mm:
```diff
@@ -26,10 +26,12 @@
 #import "config.h"
 #import "HardwareCapabilities.h"
 
+#import "WebGPUSwiftInternal.h"
```
shouldn't this be under `ENABLE(WEBGPU_SWIFT)`?

mwyrzykowski reviewed (approved):
✅ with compiler check corrected

geoffreygaren commented on Source/WebGPU/WebGPU/HardwareCapabilities.mm:
```diff
@@ -167,17 +167,71 @@ bool isShaderValidationEnabled(id<MTLDevice> device)
+    if (clangVersion[0] == 1700 && clangVersion[1] >= 6 && clangVersion[2] >= 1 && clangVersion[3] >= 1)
+        return true;
+
+    if (clangVersion[0] >= 2100 && clangVersion[1] >= 0 && clangVersion[2] >= 101 && clangVersion[3] >= 15)
```
I think this wants to be `== 2100`, and then you want a separate check for just plain `>= 2100` with no sub-versions. Otherwise the feature gets disable on the bump to 2200.0.0.0.

tadeuzagallo committed:
[WebGPU Swift] Enable runtime flag by default when compiler is compatible
https://bugs.webkit.org/show_bug.cgi?id=300935
rdar://162791460

Reviewed by Mike Wyrzykowski.

Change the runtime flag that enables WebGPU Swift such that if the compiler is new
enough to support all the features we need, then it's enabled by default. It remains
possible to override it and turn it on off on any compiler version.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Capabilities.swift: Added.
(swiftCompilerSupportsWebGPU):
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::isWebGPUSwiftEnabled):

Canonical link: https://commits.webkit.org/301749@main
